File: data-structure/heap/py/heap.py
def heap_make(A, cmp = lambda a,b: a < b):
    i = len(A) / 2 - 1
    while i >= 0:
        heapify(A, len(A), i, cmp)
        i = i - 1


def heap_insert(A, heap_size, value, cmp = lambda a,b: a < b):
    A[heap_size] = value
    i = heap_size
    while has_parent(i) and not cmp(A[i], A[parent(i)]):
        A[i], A[parent(i)] = A[parent(i)], A[i]
        i = parent(i)


def heap_delete(A, heap_size, cmp = lambda a,b: a < b):
    A[0], A[heap_size-1] = A[heap_size-1], A[0]
    heapify(A, heap_size-1, 0, cmp)


# helper functions
def parent(pos):
    return (pos - 1) / 2

def left(pos):
    return 2 * pos + 1

def right(pos):
    return 2 * pos + 2

def has_parent(pos):
    return pos != 0

def has_left(pos, heap_size):
    return left(pos) < heap_size

def has_right(pos, heap_size):
    return right(pos) < heap_size

def heapify(A, heap_size, pos, cmp = lambda a,b: a < b):
    largest = pos
    if has_left(pos, heap_size) and not cmp(A[left(pos)], A[largest]):
        largest = left(pos)
    if has_right(pos, heap_size) and not cmp(A[right(pos)], A[largest]):
        largest = right(pos)
    if largest != pos:
        A[largest], A[pos] = A[pos], A[largest]
        heapify(A, heap_size, largest, cmp)

File: data_structure/heap/java/Heap.java
// Max heap.
public class Heap {
    public static void make(int[] arr) {
	for (int i = arr.length; i >= 0; i--) {
	    heapify(arr, arr.length, i);
	}
    }

    public static void insert(int[] arr, int heap_size, int value) {
	arr[heap_size] = value;
	int i = heap_size;
	while (hasParent(i) && arr[parent(i)] < arr[i]) {
	    swap(arr, i, parent(i));
	    i = parent(i);
	}
    }

    public static void delete(int[] arr, int heap_size) {
	swap(arr, 0, heap_size-1);
	heapify(arr, heap_size-1, 0);
    }

    private static int parent(int pos) {
	return (pos - 1) / 2;
    }

    private static int left(int pos) {
	return 2 * pos + 1;
    }

    private static int right(int pos) {
	return 2 * pos + 2;
    }

    private static boolean hasParent(int pos) {
	return pos != 0;
    }

    private static boolean hasLeft(int pos, int heap_size) {
	return left(pos) < heap_size;
    }

    private static boolean hasRight(int pos, int heap_size) {
	return right(pos) < heap_size;
    }

    private static void heapify(int[] arr, int heap_size, int pos) {
	int largest = pos;
	if (hasLeft(pos, heap_size) &&
	    arr[largest] < arr[left(pos)]) {
	    largest = left(pos);
	}
	if (hasRight(pos, heap_size) &&
	    arr[largest] < arr[right(pos)]) {
	    largest = right(pos);
	}
	if (largest != pos) {
	    swap(arr, largest, pos);
	    heapify(arr, heap_size, largest);
	}
    }

    private static void swap(int[] arr, int i, int j) {
	int t = arr[i];
	arr[i] = arr[j];
	arr[j] = t;
    }
}
